src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -83,39 +83,46 @@ object Task {
 
     private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
 
+    //scalastyle:off cyclomatic.complexity method.length
     override def update(update: TaskStateOp): TaskStateChange = update match {
       // case 1: now running
-      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
         val updated = copy(
           status = status.copy(
             startedAt = Some(now),
             mesosStatus = mesosStatus))
-        TaskStateChange.Update(updated)
+        TaskStateChange.Update(updated, Some(this))
 
       // case 2: terminal
-      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Terminal(_), now) =>
-        TaskStateChange.Expunge
+      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Terminal(_), now) =>
+        TaskStateChange.Expunge(this)
 
       // case 3: health or state updated
-      case TaskStateOp.MesosUpdate(taskStatus, now) =>
-        val healthOrStateChange = for {
-          current <- status.mesosStatus
-          update <- taskStatus.mesosStatus
-          newStatus <- updatedHealthOrState(current, update)
-        } yield TaskStateChange.Update(copy(status = status.copy(mesosStatus = Some(newStatus))))
-
-        healthOrStateChange.getOrElse{
+      case TaskStateOp.MesosUpdate(_, taskStatus, now) =>
+        updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus).map { newStatus =>
+          val updatedTask = copy(status = status.copy(mesosStatus = Some(newStatus)))
+          TaskStateChange.Update(updatedTask, Some(this))
+        } getOrElse {
           log.debug("Ignoring status update for {}. Status did not change.", taskId)
-          TaskStateChange.NoChange
+          TaskStateChange.NoChange(taskId)
         }
 
-      // failure case: Launch
-      case _: TaskStateOp.Launch =>
-        TaskStateChange.Failure("Unable to handle Launch op on LaunchedEphemeral")
+      case TaskStateOp.ForceExpunge(_) =>
+        TaskStateChange.Expunge(this)
 
-      // failure case: Timeout
-      case TaskStateOp.ReservationTimeout =>
-        TaskStateChange.Failure("Unable to handle Timeout op on LaunchedEphemeral")
+      // FIXME (3221): If the task needs to recreated after a not-accepted taskOp, Create is used.
+      // that's neither nice nor obvious, so this should be improved
+      case TaskStateOp.Create(task) =>
```

Comment:
And it could take an Option of a task such that it translates to an `Update` or an `Expunge`